[JavonTeo] iP

JavonTeo opened the pull request:
# Duke

> My brain is like the Bermuda Triangle. Information goes in and then it is never found again - Unknown ([source](https://www.epicquotes.com/my-brain-is-like-the-bermuda-triangle-information-goes-in-and-then-it-is-never-found-again/))

Are you a forgetful person? Ever experienced that awkward situation where you said yes to a family event, but you forgot that you had promised your boyfriend/girlfriend to bring he/her on a date on the same day, so now your mom is arguing with your partner??? Well, Duke is here for you!

Duke frees your mind of having to remember things you need to do. It's,
- **text-based**
- **easy to learn**
- **~~FAST~~ _SUPER_ FAST to use**

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉
5. And it is FREE!

Features:
- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

### It can be used either:
1. via a text-based UI, or
2. via a GUI

To use the program via the text-based UI, run the `main` method in `Duke`:
```java
public class Duke {
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

To use the program via the GUI, run the `main` method in `Launcher`:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

JavonTeo committed:
Level-1

JavonTeo committed:
Level-2

JavonTeo committed:
Level-3

JavonTeo committed:
Level-4

JavonTeo committed:
Level-5

JavonTeo committed:
Level-6

JavonTeo committed:
A-TextUiTesting

JavonTeo committed:
Edited Duke.java: Level-7 save tasks automatically

JavonTeo committed:
Level-8

JavonTeo committed:
Merge branch 'branch-Level-8'

JavonTeo committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

tituswe commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,27 @@
+        }
+        System.out.println(this);
+    }
+}
```
Have you considered implementing a `toString()` method for this parent class? This way you need not repeat the done status or the description in each of the child classes of `Task`.

tituswe commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,165 @@
+                command = s;
+            }
+            try {
+                switch (command) {
+                    case "bye" -> {
+                        Duke.bye();
+                        DukeContinues = false;
+                    }
+                    case "list" -> Duke.list(taskList);
+                    case "mark" -> {
+                        Duke.mark(taskList, restOfString);
+                        taskListChanged = true;
+                    }
+                    case "unmark" -> {
+                        Duke.unmark(taskList, restOfString);
+                        taskListChanged = true;
+                    }
+                    case "todo" -> {
+                        Duke.addTodo(taskList, restOfString);
+                        taskListChanged = true;
+                    }
+                    case "deadline" -> {
+                        Duke.addDeadline(taskList, restOfString);
+                        taskListChanged = true;
+                    }
+                    case "event" -> {
+                        Duke.addEvent(taskList, restOfString);
+                        taskListChanged = true;
+                    }
+                    case "delete" -> {
+                        Duke.delete(taskList, restOfString);
+                        taskListChanged = true;
+                    }
+                    default -> throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
```
According to the [CS2103T coding standard](https://se-education.org/guides/conventions/java/intermediate.html) switch statements should not have indentation for the `case` clauses. Slight amendment needed here 👍

tituswe commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,165 @@
+            taskList.add(t);
+            System.out.println("Now you have " + taskList.size() + " tasks in the list.");
+        }
+    }
+
+    private static void addDeadline(ArrayList<Task> taskList, String restOfString) throws DukeException {
+        if (restOfString.isBlank()) {
+            throw new DukeException("☹ OOPS!!! The description of a deadline cannot be empty.");
+        } else {
+            Deadline t = Deadline.addDeadline(restOfString);
+            taskList.add(t);
+            System.out.println("Now you have " + taskList.size() + " tasks in the list.");
+        }
+    }
+
+    private static void addEvent(ArrayList<Task> taskList, String restOfString) throws DukeException {
+        if (restOfString.isBlank()) {
+            throw new DukeException("☹ OOPS!!! The description of a event cannot be empty.");
+        } else {
+            Event t = Event.addEvent(restOfString);
+            taskList.add(t);
+            System.out.println("Now you have " + taskList.size() + " tasks in the list.");
+        }
+    }
+
+    private static void delete(ArrayList<Task> taskList, String restOfString) throws DukeException {
+        try {
+            int i = Integer.parseInt(restOfString);
+            Task removedTask = taskList.remove(i-1);
+            System.out.println("Noted. I've removed this task:");
+            System.out.println(removedTask);
+            System.out.println("Now you have " + taskList.size() + " tasks in the list.");
+        } catch (NumberFormatException err) {
+            throw new DukeException("☹ OOPS!!! " + restOfString + " is not a valid integer for indexing the task list.");
+        } catch (IndexOutOfBoundsException err) {
+            throw new DukeException("☹ OOPS!!! There are less than " + restOfString + " tasks.");
+        }
+    }
+
 }
```
You could consider using inheritance and polymorphism for your commands, this way you wouldn't have to keep adding new cases for new commands in your `main` code. On top of this your `main` class would be a lot cleaner!

tituswe commented on src/main/java/ToDo.java:
```diff
@@ -0,0 +1,17 @@
+
+    @Override
+    public String toString() {
+        return "[T]" + super.getStatusIcon() + " " + super.getDescription();
```
I like the use of the `getStatusIcon()` and `getDescription()` methods for overriding the child class `toString()` method, however you could consider implementing the `toString()` method for the `Task` class as its a more conventional way to do it and it gets the job done just as well!

tituswe reviewed (commented):
Looks good! A few coding standard issues and suggestions that you could use. All the best!

JavonTeo committed:
Added more OOP structure according to A-MoreOOP

JavonTeo committed:
A-Packages: Organize into Packages

swx0 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,55 @@
+            command = fullCommand;
+        }
+        Command c;
+        switch (command) {
+            case "bye": {
+                c = new ByeCommand();
+                break;
+            }
+            case "list": {
+                c = new ListCommand();
+                break;
+            }
+            case "mark": {
+                c = new MarkCommand(restOfString);
+                break;
+            }
+            case "unmark": {
+                c = new UnmarkCommand(restOfString);
+                break;
+            }
+            case "todo":
+            case "deadline":
+            case "event": {
+                c = new AddCommand(command, restOfString);
+                break;
+            }
+            case "delete": {
+                c = new DeleteCommand(restOfString);
+                break;
+            }
+            default:
+                throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
```
Don't need to indent for cases/default within the switch

swx0 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,38 @@
+        System.out.println(err);
+    }
+
+    public void show(String msg) {
+        System.out.println(msg);
+    }
```
Perhaps an additional class could be used to manage the logic of printing output by the various Commands. Instead of having to use the `show(String msg)` function each time an output line is displayed.

swx0 reviewed (commented):
Overall good. Just some suggestions to consider.

JavonTeo committed:
Add JUnit tests

JavonTeo committed:
A-Jar

JavonTeo committed:
Add JavaDoc comments

JavonTeo committed:
Revert "Add JavaDoc comments"

This reverts commit a6bd37d842e1319440984f205da7e46a165db4dc.

JavonTeo committed:
Add JavaDoc comments

JavonTeo committed:
A-CodingStandard

JavonTeo committed:
Level-9

JavonTeo committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/command/ByeCommand.java

JavonTeo committed:
Merge branch 'branch-Level-9'

JavonTeo committed:
Add JavaDoc for FindCommand

JavonTeo committed:
A-Checkstyle

JavonTeo committed:
Level-10: GUI
Added GUI feature to Duke, using JavaFX technology.

JavonTeo committed:
Merge branch 'branch-Level-10'

JavonTeo committed:
Several parts of the code rely on assumptions before it can work.

The assumptions may not hold in certain situations, and may lead to
error.

Adding assertions into the code will handle these errors.

Let's start adding assertions into the code in order to check if
assumptions are held when the program is running.

JavonTeo committed:
Merge pull request #2 from JavonTeo/branch-Assertions

Branch assertions

JavonTeo committed:
Certain parts of the code does not follow good code quality standards.

An improvement in code quality will greatly improve readability of the
code.

Let's write good quality code, by following coding quality standards.

JavonTeo committed:
Merge pull request #3 from JavonTeo/branch-CodeQuality

Certain parts of the code does not follow good code quality standards.

JavonTeo committed:
The description of a Task cannot be updated.

This makes it troublesome for a user to edit the description
of the Task, as the user will have to delete the Task and create a new
Task just to edit the description.

Let's create a functionality that will allow users to update the
description of a Task.

JavonTeo committed:
Added a representative screenshot of the product.
Added a User Guide to the product.
Bug fixes.

JavonTeo committed:
removed duke.jar file

JavonTeo committed:
Created a new JAR file ready to be released
